R/teal_lockfile.R

```diff
@@ -23,92 +44,140 @@
+#'
+#' @name teal_lockfile
+#' @rdname teal_lockfile
 #'
 #' @keywords internal
-teal_lockfile <- function() {
+NULL
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_invoke <- function() {
   lockfile_path <- "teal_app.lock"
+  shiny::onStop(function() file.remove(lockfile_path))
+
   # If user has setup the file, there is no need to compute a new one.
   user_lockfile <- getOption("teal.renv.lockfile", "")
   if (!identical(user_lockfile, "")) {
     if (file.exists(user_lockfile)) {
       file.copy(user_lockfile, lockfile_path)
-      return(invisible(NULL))
+      logger::log_trace('Lockfile set using option "teal.renv.lockfile" - skipping automatic creation.')
+      return(NULL)
     } else {
       stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
     }
   }
 
-  if (!(is_in_test() || is_r_cmd_check())) {
-    old_plan <- future::plan()
-    # If there is already a parallel (non-sequential) backend, reuse it.
-    if (inherits(old_plan, "sequential")) {
-      future::plan(future::multisession, workers = 2)
-    }
+  process <- ExtendedTask$new(function(...) {
+    mirai::mirai(
+      run(lockfile_path = lockfile_path, opts = opts, sysenv = sysenv, libpaths = libpaths, wd = wd),
+      ...
+    )
+  })
 
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  process$invoke(
+    lockfile_path = lockfile_path,
+    run = renv_snapshot,
+    opts = options(),
+    libpaths = .libPaths(),
+    sysenv = as.list(Sys.getenv()), # normally output is a class of "Dlist"
+    wd = getwd()
+  )
+  logger::log_trace("Lockfile creation started based on { getwd() }.")
+
+  process
 }
 
-create_renv_lockfile <- function(close = FALSE, lockfile_path = NULL) {
-  checkmate::assert_flag(close)
-  checkmate::assert_string(lockfile_path, na.ok = TRUE)
-  promise <- promises::future_promise({
-    # Below we can not use a file created in tempdir() directory.
-    # If a file is created in tempdir(), it gets deleted on 'then(onFulfilled' part.
-    shiny::onStop(function() file.remove(lockfile_path))
+#' @rdname teal_lockfile
+#' @keywords internal
+renv_snapshot <- function(lockfile_path = NULL, opts, sysenv, libpaths, wd) {
+  checkmate::assert_string(lockfile_path)
+  checkmate::assert_list(opts)
+  checkmate::assert_class(sysenv, "list")
+  checkmate::assert_character(libpaths, min.len = 1)
+  checkmate::assert_directory(wd)
 
-    renv_logs <- utils::capture.output(
-      renv::snapshot(
-        lockfile = lockfile_path,
-        prompt = FALSE,
-        force = TRUE
-        # type = is taken from renv::settings$snapshot.type()
-      )
+  # mirai starts in vanilla session in the R.home directory. We need to pass all session related info
+  options(opts)
+  lapply(names(sysenv), function(sysvar) do.call(Sys.setenv, sysenv[sysvar]))
```

Comment:
I think this is sufficient.
```suggestion
  do.call(Sys.setenv, sysenv)
```

Reply:
Good point

```r
Sys.setenv(ZZZ = "1")
vars <- Sys.getenv()
Sys.setenv(ZZZ = "2")
do.call(Sys.setenv, as.list(vars))
Sys.getenv()[["ZZZ"]] # expects 1
#> [1] "1"
```